packages/panzer/disc-fe/src/Panzer_IntegrationValues2.cpp

```diff
@@ -846,41 +774,51 @@ generateSurfaceCubatureValues(const PHX::MDField<Scalar,Cell,NODE,Dim>& in_node_
+        for (size_type j = 0; j < contravarient_k.extent(3); ++j)
+          covarient_k(cell,ip,i,j) = 0.0;
 
       // g^{ij} = \frac{\parital x_i}{\partial \chi_\alpha}\frac{\parital x_j}{\partial \chi_\alpha}
-      for (size_type i = 0; i < contravarient.extent(2); ++i) {
-        for (size_type j = 0; j < contravarient.extent(3); ++j) {
-          for (size_type alpha = 0; alpha < contravarient.extent(2); ++alpha) {
-            covarient(cell,ip,i,j) += jac(cell,ip,i,alpha) * jac(cell,ip,j,alpha);
+      for (size_type i = 0; i < contravarient_k.extent(2); ++i) {
+        for (size_type j = 0; j < contravarient_k.extent(3); ++j) {
+          for (size_type alpha = 0; alpha < contravarient_k.extent(2); ++alpha) {
+            covarient_k(cell,ip,i,j) += jac_k(cell,ip,i,alpha) * jac_k(cell,ip,j,alpha);
           }
         }
       }
-
-    }
+    });
+    PHX::Device::execution_space().fence();
   }
 
-  auto s_contravarient = Kokkos::subview(contravarient.get_view(), std::make_pair(0,num_cells),Kokkos::ALL,Kokkos::ALL,Kokkos::ALL);
-  auto s_covarient = Kokkos::subview(covarient.get_view(), std::make_pair(0,num_cells),Kokkos::ALL,Kokkos::ALL,Kokkos::ALL);
-  Intrepid2::RealSpaceTools<PHX::Device::execution_space>::inverse(s_contravarient, s_covarient);
+  {
+    auto s_contravarient = Kokkos::subview(contravarient.get_view(), std::make_pair(0,num_cells),Kokkos::ALL,Kokkos::ALL,Kokkos::ALL);
+    auto s_covarient = Kokkos::subview(covarient.get_view(), std::make_pair(0,num_cells),Kokkos::ALL,Kokkos::ALL,Kokkos::ALL);
+    Intrepid2::RealSpaceTools<PHX::Device::execution_space>::inverse(s_contravarient, s_covarient);
+    PHX::Device::execution_space().fence();    
+  }
 
   // norm of g_ij
-  for (int cell = 0; cell < num_cells; ++cell) {
-    for (size_type ip = 0; ip < contravarient.extent(1); ++ip) {
-      norm_contravarient(cell,ip) = 0.0;
-      for (size_type i = 0; i < contravarient.extent(2); ++i) {
-        for (size_type j = 0; j < contravarient.extent(3); ++j) {
-          norm_contravarient(cell,ip) += contravarient(cell,ip,i,j) * contravarient(cell,ip,i,j);
+  {
+    auto contravarient_k = this->contravarient.get_static_view();
+    auto norm_contravarient_k = this->norm_contravarient.get_static_view();
+    Kokkos::MDRangePolicy<PHX::Device::execution_space,Kokkos::Rank<2>> policy({0,0},{(int)num_cells,(int)contravarient.extent(1)});
+    Kokkos::parallel_for("covarient metric tensor",policy,KOKKOS_LAMBDA (const int cell,const int ip) {
+      norm_contravarient_k(cell,ip) = 0.0;
+      for (size_type i = 0; i < contravarient_k.extent(2); ++i) {
+        for (size_type j = 0; j < contravarient_k.extent(3); ++j) {
+          norm_contravarient_k(cell,ip) += contravarient_k(cell,ip,i,j) * contravarient_k(cell,ip,i,j);
         }
       }
-      norm_contravarient(cell,ip) = std::sqrt(norm_contravarient(cell,ip));
-    }
+      norm_contravarient_k(cell,ip) = std::sqrt(norm_contravarient_k(cell,ip));
```

Comment:
Can you call std::sqrt from within a kokkos functor?

Reply:
good catch. This actually works on cuda, but the preferred way that we have been using during atdm is to call transcedentals is without the std namespace. non-namespaced functions have a special version provided by cuda. I talked to Christian this morning and it looks like kokkos is now providing a math function routines for everything in the standard due to SYCL having a different namespace for it's version of the functions. So I switched everything to the Kokkos provided versions as part of this PR.